### ISSUE TITLE
Optimization: Don't use any temps for most straightforward deconstruction

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#13632]
We can do away with `V_0` and `V_1` in the IL generated below.

``` C#
        [Fact]
        public void SimpleAssignWithoutConversion()
        {
            string source = @"
class C
{
    static void Main()
    {
        int x;
        string y;

        (x, y) = new C();
        System.Console.WriteLine(x + "" "" + y);
    }

    public void Deconstruct(out int a, out string b)
    {
        a = 1;
        b = ""hello"";
    }
}
";

            var comp = CompileAndVerify(source, expectedOutput: "1 hello", additionalRefs: s_valueTupleRefs);
            comp.VerifyDiagnostics();
            comp.VerifyIL("C.Main", @"
{
  // Code size       47 (0x2f)
  .maxstack  3
  .locals init (int V_0, //x
                string V_1, //y
                int V_2,
                string V_3,
                string V_4)
  IL_0000:  newobj     ""C..ctor()""
  IL_0005:  ldloca.s   V_2
  IL_0007:  ldloca.s   V_3
  IL_0009:  call       ""void C.Deconstruct(out int, out string)""
  IL_000e:  ldloc.2
  IL_000f:  ldloc.3
  IL_0010:  stloc.s    V_4
  IL_0012:  dup
  IL_0013:  stloc.0
  IL_0014:  ldloc.s    V_4
  IL_0016:  stloc.1
  IL_0017:  pop
  IL_0018:  ldloc.0
  IL_0019:  box        ""int""
  IL_001e:  ldstr      "" ""
  IL_0023:  ldloc.1
  IL_0024:  call       ""string string.Concat(object, object, object)""
  IL_0029:  call       ""void System.Console.WriteLine(string)""
  IL_002e:  ret
}");
        }
```

**Comment by gafter** [comment]
The work to address this will probably be much easier once the phase ordering error described in #13270 has been addressed.

**Comment by jcouv** [comment]
Fixed by #17027
With the fix, many temps have been optimized away (mostly as the result of not storing the initial parts on the right in a ValueTuple and fetching them out).
